lib/event_store_client/errors.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+module EventStoreClient
+  class Error < StandardError
+    # @return [Hash]
+    def as_json(*)
+      to_h.transform_keys(&:to_s)
+    end
+
+    # @return [Hash]
+    def to_h
+      hash =
+        instance_variables.each_with_object({}) do |var, result|
+          key = var.to_s
+          key[0] = '' # remove @ sign
+          result[key.to_sym] = instance_variable_get(var)
+        end
+      hash[:message] = message
+      hash[:backtrace] = backtrace
+      hash
+    end
+  end
+
+  class StreamNotFoundError < Error
+    attr_reader :stream_name
+
+    # @param stream_name [String]
+    def initialize(stream_name)
+      @stream_name = stream_name
+      super("Stream #{stream_name.inspect} does not exist.")
+    end
+  end
+
+  class WrongExpectedVersionError < Error
+    attr_reader :wrong_expected_version, :caused_by
+
+    # @param wrong_expected_version [EventStore::Client::Streams::AppendResp::WrongExpectedVersion]
+    # @param caused_by [EventStoreClient::DeserializedEvent] an event on which
+    #   WrongExpectedVersionError error happened. It can be useful when appending array of events -
+    #   based on it you will know which events were appended and which weren't.
+    def initialize(wrong_expected_version, caused_by:)
+      @wrong_expected_version = wrong_expected_version
+      @caused_by = caused_by
+      super(user_friendly_message)
+    end
+
+    private
+
+    # @return [String]
+    def user_friendly_message
+      if wrong_expected_version.expected_stream_exists
```

Comment:
I believe it is possiblie to use some mapping here like:
```
messages = { 
  expected_stream_exists: 'Expected stream to exist, but it doesn't.',
  expected_no_stream: 'Expected stream to be absent, but it actually exists.'
  ....
}
messages[wrong_expected_version.key]
```

Just an idea we can discuss. But my idea is to avoid the tone of if and if you need to add another message you have to add another if in this approach. So with number of if the whole lookup slows down

Reply:
The condition is a bit more complex than look up of one attribute.

Reply:
Anyway I have a feeling we can try to refactor it as those ifs looks not the best to me.